### PR TITLE
refactor: harden release 3 code quality

### PR DESCRIPTION
**Pull request by kantorcodes**
## Summary

This change performs a release-wide code-quality audit and a behavior-preserving remediation pass ahead of HOL Guard 3.0.

The audit records the current structural-debt baseline without treating size or complexity alone as a defect. It covers oversized files, long and high-complexity Python functions, exact function copies, silent broad exception handlers, parse failures, and temporary delivery residue.

## What changed

- Added a deterministic AST-based audit, committed baseline, report generator, and CI ratchet. New pull requests may reduce debt but cannot silently introduce or grow the audited debt classes.
- Consolidated repeated implementation patterns into focused shared modules for JSON loading, hook payloads, state files, workspace overrides, artifact and file identity, collection coercion, stable JSON, private and durable file I/O, path security, contained execution, command candidates, package evidence, executable resolution, Node semver handling, time handling, and MDM platform support.
- Replaced copy-pasted adapter, CLI, runtime, containment, package-evidence, MDM, and storage helpers with shared implementations while preserving existing contracts and call-site behavior.
- Hardened edge cases around atomic writes, private-file permissions, malformed state, platform-specific paths, managed-install detection, executable resolution, version parsing, containment cleanup, bounded payload handling, and shell-option parsing.
- Added regression coverage for the audit, durable and private file I/O, managed-install context, Node semver behavior, and Kubernetes shell-command parsing.
- Removed obsolete temporary delivery payloads and one-shot shepherd workflows.

## Compatibility and security

- No public API removals or dependency additions.
- Security-sensitive behavior remains fail-closed where it was already fail-closed.
- Shared helpers preserve existing defaults, return shapes, and exception semantics unless the previous behavior was an identified unsafe or malformed-input edge case.
- Private-file helpers retain no-follow, ownership, permission, bounded-read, and metadata-stability checks.
- Generated assets and lockfiles remain classified separately so the ratchet does not encourage unsafe manual edits.

## Validation

- Structural audit and baseline verification
- Ruff lint and format checks
- BasedPyright type checking
- Python compilation checks
- Source-bound command decision-diff verification
- Focused regression tests for every changed edge case
- Full repository CI before merge

### CODE REVIEW

**Comment by kantorcodes** [comment]
@coderabbitai review

File: src/codex_plugin_scanner/guard/runtime/kubernetes_command_support.py [hunk 1]
```diff
@@ -82,6 +82,47 @@ def kubernetes_option_tokens_consumed(
...
+
+def shell_command_script(tokens: tuple[str, ...]) -> str | None:
+    """Return the script passed to a shell's ``-c`` option, if present."""
+
+    index = 0
+    while index < len(tokens):
+        token = tokens[index]
+        if token == "-c" and index + 1 < len(tokens):
+            return tokens[index + 1]
+        if token.startswith("-") and "c" in token[1:] and index + 1 < len(tokens):
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply src/codex_plugin_scanner/guard/runtime/kubernetes_command_support.py]: Fixed. `shell_command_script` now explicitly rejects tokens beginning with `--` before checking short-option clusters, so long options such as `--context` and `--config` cannot consume the following value as a shell script. The focused Kubernetes command-support tests pass.

File: src/codex_plugin_scanner/guard/private_file_io.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        return None
+    try:
+        opened = os.fstat(descriptor)
+        if not _private_file_metadata_is_valid(opened) or not _same_file(path_before, opened):
+            return None
+        if opened.st_size > max_bytes:
+            return None
+        chunks: list[bytes] = []
+        consumed = 0
+        while consumed <= max_bytes:
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply src/codex_plugin_scanner/guard/private_file_io.py]: Fixed. The bounded read loop now uses `while consumed < max_bytes` and reads exactly `max_bytes - consumed` bytes at most, with no extra-byte request.

File: scripts/ci/code_quality_audit.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+        ".yml",
+    }
+)
+EXCLUDED_PARTS = frozenset(
+    {
+        ".git",
+        ".mypy_cache",
+        ".pytest_cache",
+        ".ruff_cache",
+        ".venv",
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply scripts/ci/code_quality_audit.py:64]: Fixed. `EXCLUDED_PARTS` now includes both `.venv` and plain `venv`, preventing developer-created virtual environments from affecting the audit.

File: scripts/ci/code_quality_audit.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+            if not path.is_file() or path.suffix.lower() not in CODE_SUFFIXES:
+                continue
+            relative = path.relative_to(root)
+            if any(part in EXCLUDED_PARTS for part in relative.parts) or relative in seen:
+                continue
+            seen.add(relative)
+            yield path
+
+
+def _category(path: str) -> str:
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply scripts/ci/code_quality_audit.py:110]: Fixed. Test categorization now recognizes common test directories (`test`, `tests`, `testing`, `spec`, `specs`, and `__tests__`) plus `test_*`, `*_test.py`, `*_tests.py`, and standard `.test.`/`.spec.` filename patterns.

File: scripts/ci/code_quality_audit.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    expected_duplicates = {
+        str(name): int(value) for name, value in baseline.get("duplicate_function_groups", {}).items()
+    }
+    for group in report["duplicate_function_groups"]:
+        digest = str(group["digest"])
+        count = len(group["instances"])
+        if digest not in expected_duplicates:
+            locations = ", ".join(f"{item['path']}:{item['line']}" for item in group["instances"])
+            failures.append(f"Duplicate function group introduced ({count} instances): {locations}")
+        elif count > expected_duplicates[digest]:
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply scripts/ci/code_quality_audit.py]: Fixed. The duplicate-function ratchet now compares the exact current and baseline location sets for every digest. Any introduced location fails, and any removed location marks the baseline stale, so lower-count reintroduction cannot pass silently.

File: scripts/ci/code_quality_audit.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+            failures.append(f"Duplicate function group grew: {digest} ({expected_duplicates[digest]} -> {count})")
+    expected_handlers = Counter(
+        {str(name): int(value) for name, value in baseline.get("silent_broad_exception_handlers", {}).items()}
+    )
+    current_handlers = Counter(str(item["identity"]) for item in report["silent_broad_exception_handlers"])
+    for identity, count in current_handlers.items():
+        if count > expected_handlers[identity]:
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply scripts/ci/code_quality_audit.py]: Fixed. Silent broad-exception handlers are now compared as exact identity-and-line sets. Newly introduced handlers fail, while removed handlers require a baseline refresh, preventing lower-count reintroduction from passing.

File: scripts/ci/code_quality_ast.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        self.stack: list[str] = []
+        self.functions: list[FunctionMetric] = []
+        self.handlers: list[SilentHandler] = []
+
+    def visit_ClassDef(self, node: ast.ClassDef) -> None:
+        self.stack.append(node.name)
+        self.generic_visit(node)
+        self.stack.pop()
+
+    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply scripts/ci/code_quality_ast.py:138]: Fixed. `_FunctionCollector.visit_Lambda` now records lambdas with stable source-position qualnames and includes them in line, complexity, digest, and nested-handler analysis.

File: scripts/ci/code_quality_ast.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+        for handler in handler_visitor.handlers:
+            exception = _exception_name(handler.type)
+            if exception in {"bare", "BaseException", "Exception"}:
+                self.handlers.append(SilentHandler(self.path, qualname, handler.lineno, exception))
+        self.generic_visit(node)
+        self.stack.pop()
+
+
+def _canonical_ast(value: object) -> object:
+    """Return a stable, version-independent representation of an AST value."""
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply scripts/ci/code_quality_ast.py]: Fixed. `_canonical_ast` now sorts `ast.iter_fields` by field name before canonicalization, removing dependence on interpreter field-order differences.

File: scripts/ci/code_quality_ast.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    return True
+
+
+def _exception_name(node: ast.expr | None) -> str:
+    if node is None:
+        return "bare"
+    if isinstance(node, ast.Name):
+        return node.id
+    if isinstance(node, ast.Attribute):
+        return node.attr
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply scripts/ci/code_quality_ast.py]: Fixed. `_exception_name` now recursively preserves attribute prefixes, so identities such as `foo.Bar` and `baz.Bar` remain distinct while broad-exception detection still checks the terminal class name.

File: docs/guard/release-3-code-quality-hardening-prd.md [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+| Production/tooling Python functions | 11,578 | 11,534 | -44 |
+| Python functions at least 100 lines | 247 | 247 | unchanged and ratcheted |
+| Python functions at complexity 30+ | 154 | 154 | unchanged and ratcheted |
+| Exact Python function-copy groups | 48 | 1 | -47 |
+| Functions in exact-copy groups | 104 | 2 | -102 |
+| Silent broad exception handlers | 36 | 36 | unchanged and ratcheted |
+| Forbidden one-shot delivery artifacts | 18 | 0 | -18 |
+| Python parse errors | 0 | 0 | unchanged |
+
+The one remaining exact-copy group is intentional. `scripts/mdm/generate-release-manifest.py` is a standalone release tool that must retain a local no-follow reader instead of importing the installed HOL Guard package. The production copy remains in `guard/mdm/manifest.py`. Both are security-sensitive and covered by tests.
```
[the remark this answers is not in the document]
Comment (kantorcodes) [reply docs/guard/release-3-code-quality-hardening-prd.md]: Addressed by removing the internal planning PRD from the public repository diff. The retained public audit documentation uses repository-resolvable paths, while private planning artifacts are now explicitly rejected by the final hygiene check.
